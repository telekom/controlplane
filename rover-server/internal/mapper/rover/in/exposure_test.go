// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package in

import (
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	roverv1 "github.com/telekom/controlplane/rover/api/v1"

	"github.com/telekom/controlplane/rover-server/internal/api"
)

var _ = Describe("Exposure Mapper", func() {
	Context("mapApiExposure", func() {
		It("must map ApiExposure correctly", func() {
			input := apiExposure

			output := mapApiExposure(input)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must handle empty ApiExposure input", func() {
			input := api.ApiExposure{}

			output := mapApiExposure(input)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("mapExposure", func() {
		It("must map an ApiExposure correctly", func() {
			input := GetApiExposure(apiExposure)
			output := &roverv1.Exposure{}

			err := mapExposure(&input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must map an EventExposure correctly", func() {
			input := GetEventExposure(eventExposure)
			output := &roverv1.Exposure{}

			err := mapExposure(&input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must return an error for unknown exposure type", func() {
			input := &api.Exposure{}
			output := &roverv1.Exposure{}

			err := mapExposure(input, output)

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("failed to get exposure type"))
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("toRoverVisibility", func() {
		It("must map WORLD visibility correctly", func() {
			input := api.WORLD

			output := toRoverVisibility(input)

			Expect(output).To(Equal(roverv1.VisibilityWorld))
		})

		It("must map ZONE visibility correctly", func() {
			input := api.ZONE

			output := toRoverVisibility(input)

			Expect(output).To(Equal(roverv1.VisibilityZone))
		})

		It("must map ENTERPRISE visibility correctly", func() {
			input := api.ENTERPRISE

			output := toRoverVisibility(input)

			Expect(output).To(Equal(roverv1.VisibilityEnterprise))
		})

		It("must map unknown visibility", func() {
			input := api.Visibility("unknown")

			output := toRoverVisibility(input)

			Expect(output).To(Equal(roverv1.Visibility("Unknown")))
		})
	})

	Context("toRoverApprovalStrategy", func() {
		It("must map AUTO approval strategy correctly", func() {
			input := api.AUTO

			output := toRoverApprovalStrategy(input)

			Expect(output).To(Equal(roverv1.ApprovalStrategyAuto))
		})

		It("must map MANUAL approval strategy correctly", func() {
			input := api.SIMPLE

			output := toRoverApprovalStrategy(input)

			Expect(output).To(Equal(roverv1.ApprovalStrategySimple))
		})

		It("must map FOUREYES approval strategy correctly", func() {
			input := api.FOUREYES

			output := toRoverApprovalStrategy(input)

			Expect(output).To(Equal(roverv1.ApprovalStrategyFourEyes))
		})

		It("must map unknown approval strategy", func() {
			input := api.ApprovalStrategy("unknown")

			output := toRoverApprovalStrategy(input)

			Expect(output).To(Equal(roverv1.ApprovalStrategy("Unknown")))
		})
	})
})
