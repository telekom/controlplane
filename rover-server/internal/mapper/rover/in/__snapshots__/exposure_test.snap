
[Exposure Mapper mapApiExposure must map ApiExposure correctly - 1]
&v1.ApiExposure{
    BasePath:  "/eni/distr/v1",
    Upstreams: {
        {URL:"https://httpbin.org/anything", Weight:0},
    },
    Visibility: "World",
    Approval:   v1.Approval{
        Strategy:     "Simple",
        TrustedTeams: nil,
    },
    Transformation: (*v1.Transformation)(nil),
    Traffic:        v1.Traffic{},
    Security:       (*v1.Security)(nil),
}
---

[Exposure Mapper mapApiExposure must handle empty ApiExposure input - 1]
&v1.ApiExposure{
    BasePath:  "",
    Upstreams: {
    },
    Visibility:     "",
    Approval:       v1.Approval{},
    Transformation: (*v1.Transformation)(nil),
    Traffic:        v1.Traffic{},
    Security:       (*v1.Security)(nil),
}
---

[Exposure Mapper mapExposure must map an ApiExposure correctly - 1]
&v1.Exposure{
    Api: &v1.ApiExposure{
        BasePath:  "/eni/distr/v1",
        Upstreams: {
            {URL:"https://httpbin.org/anything", Weight:0},
        },
        Visibility: "World",
        Approval:   v1.Approval{
            Strategy:     "Simple",
            TrustedTeams: nil,
        },
        Transformation: (*v1.Transformation)(nil),
        Traffic:        v1.Traffic{},
        Security:       (*v1.Security)(nil),
    },
    Event: (*v1.EventExposure)(nil),
}
---

[Exposure Mapper mapExposure must map an EventExposure correctly - 1]
&v1.Exposure{
    Api:   (*v1.ApiExposure)(nil),
    Event: &v1.EventExposure{EventType:"tardis.horizon.demo.cetus.v1"},
}
---

[Exposure Mapper mapExposure must return an error for unknown exposure type - 1]
&v1.Exposure{}
---
